File: python_bot/src/order_service.py
import logging
from typing import List
from .models import OrderRequest, OrderResponse

logger = logging.getLogger(__name__)


class OrderService:
    """Wraps Wealthsimple order endpoints."""

    def __init__(self, client):
        self._client = client

    def place_order(self, order: OrderRequest) -> OrderResponse:
        logger.info(
            "Placing order: %s %d x %s @ $%s",
            order.order_type.value, order.quantity, order.symbol, order.limit_price,
        )
        data = self._client.post("/orders", order.to_api_payload())
        return self._parse_order(data)

    def get_orders(self) -> List[OrderResponse]:
        data = self._client.get("/orders")
        results = data.get("results", [])
        return [self._parse_order(raw) for raw in results]

    def cancel_order(self, order_id: str) -> None:
        logger.info("Cancelling order: %s", order_id)
        self._client.delete(f"/orders/{order_id}")

    def get_pending_orders(self) -> List[OrderResponse]:
        orders = self.get_orders()
        pending = [o for o in orders if o.status in ("submitted", "pending", "new")]
        logger.info("Found %d pending orders", len(pending))
        return pending

    def _parse_order(self, raw: dict) -> OrderResponse:
        return OrderResponse(
            order_id=raw.get("order_id", raw.get("id", "")),
            security_id=raw.get("security_id", ""),
            symbol=raw.get("symbol", ""),
            quantity=int(raw.get("quantity", 0)),
            order_type=raw.get("order_type", ""),
            status=raw.get("status", ""),
            limit_price=raw.get("limit_price", {}).get("amount")
            if isinstance(raw.get("limit_price"), dict)
            else raw.get("limit_price"),
            filled_at=raw.get("filled_at"),
            created_at=raw.get("created_at"),
        )

File: dotnet_bot/src/WealthsimpleTradingBot/Api/OrderService.cs
using System.Text.Json;
using WealthsimpleTradingBot.Models;

namespace WealthsimpleTradingBot.Api;

public interface IOrderService
{
    Task<OrderResponse> PlaceOrderAsync(OrderRequest order);
    Task<List<OrderResponse>> GetOrdersAsync();
    Task CancelOrderAsync(string orderId);
    Task<List<OrderResponse>> GetPendingOrdersAsync();
}

public class OrderService : IOrderService
{
    private readonly IWealthsimpleClient _client;
    private readonly ILogger<OrderService> _logger;

    public OrderService(IWealthsimpleClient client, ILogger<OrderService> logger)
    {
        _client = client;
        _logger = logger;
    }

    public async Task<OrderResponse> PlaceOrderAsync(OrderRequest order)
    {
        _logger.LogInformation("Placing order: {Type} {Qty}x {Symbol} @ ${Price}",
            order.Type, order.Quantity, order.Symbol, order.LimitPrice);

        var data = await _client.PostAsync<JsonElement>("/orders", order.ToApiPayload());
        return ParseOrder(data);
    }

    public async Task<List<OrderResponse>> GetOrdersAsync()
    {
        var data = await _client.GetAsync<JsonElement>("/orders");
        var orders = new List<OrderResponse>();

        if (data.TryGetProperty("results", out var results))
        {
            foreach (var raw in results.EnumerateArray())
                orders.Add(ParseOrder(raw));
        }

        return orders;
    }

    public async Task CancelOrderAsync(string orderId)
    {
        _logger.LogInformation("Cancelling order: {OrderId}", orderId);
        await _client.DeleteAsync($"/orders/{orderId}");
    }

    public async Task<List<OrderResponse>> GetPendingOrdersAsync()
    {
        var orders = await GetOrdersAsync();
        var pending = orders.Where(o =>
            o.Status is "submitted" or "pending" or "new").ToList();
        _logger.LogInformation("Found {Count} pending orders", pending.Count);
        return pending;
    }

    private static OrderResponse ParseOrder(JsonElement raw)
    {
        decimal? limitPrice = null;
        if (raw.TryGetProperty("limit_price", out var lp))
        {
            if (lp.ValueKind == JsonValueKind.Object &&
                lp.TryGetProperty("amount", out var lpAmount))
                limitPrice = lpAmount.GetDecimal();
            else if (lp.ValueKind == JsonValueKind.Number)
                limitPrice = lp.GetDecimal();
        }

        return new OrderResponse(
            OrderId: GetStr(raw, "order_id", GetStr(raw, "id")),
            SecurityId: GetStr(raw, "security_id"),
            Symbol: GetStr(raw, "symbol"),
            Quantity: raw.TryGetProperty("quantity", out var q) ? q.GetInt32() : 0,
            OrderType: GetStr(raw, "order_type"),
            Status: GetStr(raw, "status"),
            LimitPrice: limitPrice,
            FilledAt: raw.TryGetProperty("filled_at", out var fa) && fa.ValueKind != JsonValueKind.Null
                ? fa.GetDateTime() : null,
            CreatedAt: raw.TryGetProperty("created_at", out var ca) && ca.ValueKind != JsonValueKind.Null
                ? ca.GetDateTime() : null
        );
    }

    private static string GetStr(JsonElement el, string name, string def = "")
        => el.TryGetProperty(name, out var val) ? val.GetString() ?? def : def;
}
